File: server/src/convert/memory-dialects.ts
import type { MemoryScope } from "../../libs/canonical/src/generated.js";

/**
 * Which tool's dialect a captured instruction file is written in.
 *
 * Every id here is a `readers` entry the capture agent stamps onto a memory
 * document, which is the only thing that ties a file to a tool. The list is
 * held to `memoar-connectors/src/memory_files.rs` by a test, because a tool
 * spelled one way in Rust and another way here is the shape of bug that has
 * already cost this repository a broken `memoar login` and a batch of inert API
 * keys.
 */
export const MEMORY_SOURCE_TOOLS = [
  "antigravity-cli",
  "claude-code",
  "codex",
  "copilot",
  "crush",
  "cursor",
  "goose",
  "kilo",
  "opencode",
  "roo",
  "zed",
] as const;

export type MemorySourceTool = (typeof MEMORY_SOURCE_TOOLS)[number];

/**
 * The dialects a conversion may be written *into*.
 *
 * `cursor` reads instructions only from `.cursor/rules/**\/*.mdc`, and Cursor
 * ignores a file there that carries no frontmatter — the connector table says
 * so in as many words. Writing one would mean inventing frontmatter, which the
 * decision behind this feature rules out; writing one without frontmatter would
 * put the user's words somewhere Cursor silently never reads, which is worse
 * than refusing. So Cursor is a source and not a target, and that is the one
 * asymmetry in this table.
 */
export const MEMORY_DIALECTS = [
  "antigravity-cli",
  "claude-code",
  "codex",
  "copilot",
  "crush",
  "goose",
  "kilo",
  "opencode",
  "roo",
  "zed",
] as const;

export type MemoryDialect = (typeof MEMORY_DIALECTS)[number];

export function isMemoryDialect(value: string): value is MemoryDialect {
  return (MEMORY_DIALECTS as readonly string[]).includes(value);
}

export interface MemoryDestination {
  /**
   * `file` — the dialect reads exactly one path, so every selected document
   * lands in it and several of them concatenate.
   *
   * `directory` — the dialect reads a rules directory, so each document keeps
   * its own file and nothing is concatenated.
   */
  kind: "file" | "directory";
  /** Relative to the root the scope names: the home directory, or the workspace. */
  path: string;
}

/**
 * Where each dialect reads, per scope. Taken from the connector table and held
 * to it by a test: memoar only ever writes where its own capture already looks,
 * so a converted file is picked up on the next reading rather than becoming a
 * copy nobody tracks.
 *
 * A scope missing from a dialect means that tool has no such file. Copilot,
 * Goose and Zed document no user-wide path; Roo and Kilo read the same rules
 * directory in both scopes.
 */
export const MEMORY_DIALECT_PATHS: Record<MemoryDialect, Partial<Record<MemoryScope, MemoryDestination>>> = {
  "antigravity-cli": {
    global: { kind: "file", path: ".gemini/GEMINI.md" },
    project: { kind: "file", path: "GEMINI.md" },
  },
  "claude-code": {
    global: { kind: "file", path: ".claude/CLAUDE.md" },
    project: { kind: "file", path: "CLAUDE.md" },
  },
  codex: {
    global: { kind: "file", path: ".codex/AGENTS.md" },
    project: { kind: "file", path: "AGENTS.md" },
  },
  copilot: {
    project: { kind: "file", path: ".github/copilot-instructions.md" },
  },
  crush: {
    global: { kind: "file", path: ".config/crush/CRUSH.md" },
    project: { kind: "file", path: "AGENTS.md" },
  },
  goose: {
    project: { kind: "file", path: ".goosehints" },
  },
  kilo: {
    global: { kind: "directory", path: ".kilocode/rules" },
    project: { kind: "directory", path: ".kilocode/rules" },
  },
  opencode: {
    global: { kind: "file", path: ".config/opencode/AGENTS.md" },
    project: { kind: "file", path: "AGENTS.md" },
  },
  roo: {
    global: { kind: "directory", path: ".roo/rules" },
    project: { kind: "directory", path: ".roo/rules" },
  },
  zed: {
    project: { kind: "file", path: ".rules" },
  },
};

export function memoryDestination(dialect: MemoryDialect, scope: MemoryScope): MemoryDestination | undefined {
  return MEMORY_DIALECT_PATHS[dialect][scope];
}

/**
 * The prefix a wire path carries so the root it is relative to is never
 * guessed. `~/` is the home directory, `./` is the workspace the conversion
 * named — the same two roots the scopes mean.
 */
export function rootPrefix(scope: MemoryScope): "~/" | "./" {
  return scope === "global" ? "~/" : "./";
}

/**
 * The maximum length of a generated rules filename.
 *
 * ext4 and APFS both stop at 255 bytes for one component, and a global memory
 * file's path can be much longer than that once the directories are flattened
 * into it. The tail is kept rather than the head because the tail is the part
 * that names the file.
 */
const MAX_RULES_NAME = 120;

/**
 * The filename one source document takes inside a rules directory.
 *
 * Derived from the source path alone, so the same file converts to the same
 * name every time and a second conversion is a no-op rather than a second copy
 * under a new name. Restricted to `[a-z0-9-]` and one `.md`, which is also the
 * shape the materializer will accept for a path inside a rules directory.
 */
export function rulesFileName(sourcePath: string): string {
  const slug = sourcePath
    .toLowerCase()
    .replaceAll(/[^a-z0-9]+/gu, "-")
    .replaceAll(/^-+|-+$/gu, "");
  const trimmed = slug.slice(-MAX_RULES_NAME).replaceAll(/^-+/gu, "");
  return `${trimmed || "memory"}.md`;
}

File: agent/crates/memoar-materializer/src/memory_dialects.rs
//! Which instruction file each tool reads, and where a conversion may write it.
//!
//! The mirror of `server/src/convert/memory-dialects.ts`, and held to it by a
//! test that reads both. Two tables in two languages with nothing comparing
//! them is the shape of bug that has broken `memoar login` in CI and made every
//! browser-minted API key inert — twice in one day. Here the cost of a
//! disagreement would be a file written where the target tool never looks.

use serde::{Deserialize, Serialize};

use crate::error::MaterializeError;
use memoar_canonical::MemoryScope;

/// A dialect a conversion can be written into.
///
/// Cursor is deliberately absent. It reads nothing from a `.cursor/rules/*.mdc`
/// that carries no frontmatter, and inventing frontmatter is not a mechanical
/// port, so Cursor is a source and never a target.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum MemoryDialect {
    AntigravityCli,
    ClaudeCode,
    Codex,
    Copilot,
    Crush,
    Goose,
    Kilo,
    Opencode,
    Roo,
    Zed,
}

impl MemoryDialect {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::AntigravityCli => "antigravity-cli",
            Self::ClaudeCode => "claude-code",
            Self::Codex => "codex",
            Self::Copilot => "copilot",
            Self::Crush => "crush",
            Self::Goose => "goose",
            Self::Kilo => "kilo",
            Self::Opencode => "opencode",
            Self::Roo => "roo",
            Self::Zed => "zed",
        }
    }

    /// Every dialect, so a caller can list them without repeating the table.
    pub const ALL: [Self; 10] = [
        Self::AntigravityCli,
        Self::ClaudeCode,
        Self::Codex,
        Self::Copilot,
        Self::Crush,
        Self::Goose,
        Self::Kilo,
        Self::Opencode,
        Self::Roo,
        Self::Zed,
    ];
}

impl std::str::FromStr for MemoryDialect {
    type Err = MaterializeError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Self::ALL
            .into_iter()
            .find(|dialect| dialect.as_str() == value)
            .ok_or_else(|| MaterializeError::UnsupportedTarget(value.to_owned()))
    }
}

/// Where a dialect reads: one exact path, or a directory of rules files.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MemoryDestination {
    /// The tool reads exactly this path, and nothing else.
    File(&'static str),
    /// The tool reads every `.md` directly inside this directory.
    Directory(&'static str),
}

/// The whole table. A `None` means that tool documents no such file.
#[must_use]
pub fn memory_destination(
    dialect: MemoryDialect,
    scope: &MemoryScope,
) -> Option<MemoryDestination> {
    use MemoryDestination::{Directory, File};
    let global = matches!(scope, MemoryScope::Global);
    match (dialect, global) {
        (MemoryDialect::AntigravityCli, true) => Some(File(".gemini/GEMINI.md")),
        (MemoryDialect::AntigravityCli, false) => Some(File("GEMINI.md")),
        (MemoryDialect::ClaudeCode, true) => Some(File(".claude/CLAUDE.md")),
        (MemoryDialect::ClaudeCode, false) => Some(File("CLAUDE.md")),
        (MemoryDialect::Codex, true) => Some(File(".codex/AGENTS.md")),
        (MemoryDialect::Codex, false) => Some(File("AGENTS.md")),
        (MemoryDialect::Copilot, true) => None,
        (MemoryDialect::Copilot, false) => Some(File(".github/copilot-instructions.md")),
        (MemoryDialect::Crush, true) => Some(File(".config/crush/CRUSH.md")),
        (MemoryDialect::Crush, false) => Some(File("AGENTS.md")),
        (MemoryDialect::Goose, true) => None,
        (MemoryDialect::Goose, false) => Some(File(".goosehints")),
        (MemoryDialect::Kilo, _) => Some(Directory(".kilocode/rules")),
        (MemoryDialect::Opencode, true) => Some(File(".config/opencode/AGENTS.md")),
        (MemoryDialect::Opencode, false) => Some(File("AGENTS.md")),
        (MemoryDialect::Roo, _) => Some(Directory(".roo/rules")),
        (MemoryDialect::Zed, true) => None,
        (MemoryDialect::Zed, false) => Some(File(".rules")),
    }
}

/// The prefix a wire path carries so the root is never guessed: `~/` is the
/// home directory, `./` is the workspace the conversion named.
#[must_use]
pub const fn root_prefix(scope: &MemoryScope) -> &'static str {
    match scope {
        MemoryScope::Global => "~/",
        MemoryScope::Project => "./",
    }
}
